### ISSUE TITLE
Make Frame enum public

### ISSUE THREAD

**Issue by mickel8** [cloudflare/quiche#485]
Hi,
is it possible to export Frame enum as public? It would be really nice to have ability to decode QUIC packets and see how they are really look like e.g. in order to better protocol understanding.

I mean this function
https://github.com/cloudflare/quiche/blob/c726d349120ce72e8f0a00cd5176fd0767325511/src/frame.rs#L144-L146

This line in src/lib.rs will do it
```rust
pub use crate::frame::Frame;
```
but I suppose there is a reason to not doing this :/

**Comment by mickel8** [comment]
I haven't noticed that there is already function `decode_pkt` but I cannot figure out when I should use it. I mean I could only use this function while receiving  message from peer but not during connection establishment. Anyway is there a possibility to see what I am passing to `conn.recv()` function? I am talking about buf content e.g.
```rust
        let len = match socket.recv(&mut buf) {
            Ok(v) => v,
            Err(e) => {
                if e.kind() == std::io::ErrorKind::WouldBlock {
                    debug!("recv() would block");
                    break;
                }
                panic!("recv() failed: {:?}", e);
            }
        };
        debug!("got {} bytes", len);

        let pkt_buf = &mut buf[..len];

        // Parse the QUIC packet's header.
        let hdr = match quiche::Header::from_slice(
            pkt_buf,
            quiche::MAX_CONN_ID_LEN,
        ) {
            Ok(v) => v,
            Err(e) => {
                error!("Parsing packet header failed: {:?}", e);
                panic!();
            },
        };
```
is there any way to see whole content of buf not only header? 

I tried this (after some lib modifications):
```rust
let mut b = quiche::Octets::with_slice(&mut buf[..len]);
info!("frame {:#?}", quiche::Frame::from_bytes(&mut b, hdr.ty));
```

and this

```rust
    let frames = decode_pkt(conn, &mut buf, len);
            match frames {
                Ok(frames) => {
                    for frame in frames {
                        info!("frame: {:?}", frame);
                    }
                }
                Err(e) => info!("{}", e)
            }
```

but the first one gives me `InvalidFrame` error and the second one `CryptoFail`. I suppose that `decode_pkt` is intended to use only after connection establishment especially after TLS handshake i.e. when we have generated  key used to encoding and decoding packets.

**Comment by LPardue** [comment]
I think you're looking at `quiche::testing::decode_pkt()`, which is a library-internal method used for unit testing and probably not going to work as you hope.

More generally, the quiche API is intended to take responsibility for the protocol aspects and therefore purposefully provides the abstractions. I appreciate the learning purposes you might have but can I ask why you want to use the quiche library to do so? A simpler is to analyse logs, or to dump the session key and dissect packet captures in Wireshark.

**Comment by mickel8** [comment]
I wanted to better understand how the library works especially while sending packets. For example
```
[2020-05-01T16:40:32Z DEBUG quicson] got 70 bytes
[2020-05-01T16:40:32Z DEBUG quicson] processed 70 bytes
[2020-05-01T16:40:32Z DEBUG quicson] got 455 bytes
[2020-05-01T16:40:32Z DEBUG quicson] processed 455 bytes
```
here `quiche` (`quicson` is my project name) received 70bytes and then 455 bytes. I was wondering why it is not in one packet when there was nothing sent between 70 and 455 bytes so I thought let's print these packets and see how they are really look like but it wasn't too easy. It seems that wireshark will be the best way, thanks!

**Comment by LPardue** [comment]
From the sound of it, the logging should fit your needs and is a bit easier than worrying about pcaps. So in this case I would recommend two logging options that can be combined:

1) Run the application with `RUST_LOG=trace` which will give you a lot more detail from the internals of quiche. For example, it will tell you QUIC packet receive and transmits and details of all the frames inside each packet

2) If you run `quiche-client` or `quiche-server` with `QLOGDIR=/path/to/folder` it will log packet and frame events to a file in the qlog format that you can visualise in a tool such as qvis https://qvis.edm.uhasselt.be/#/files

**Comment by mickel8** [comment]
Thanks a lot, everything works! Now I am going to analyze it.

**Comment by LPardue** [comment]
great, glad you can make some progress.
